src/js/relay.firefox.com/get_profile_data.js

```diff
@@ -87,38 +104,59 @@
       },
     });
 
-    // Get the relay address objects from the addon storage
-    const addonStorageRelayAddresses = await browser.storage.local.get(
-      "relayAddresses"
-    );
-
     const siteStorageEnabled = serverProfileData[0].server_storage;
 
-    const addonRelayAddresses =
-      Object.keys(addonStorageRelayAddresses).length === 0
-        ? { relayAddresses: [] }
-        : addonStorageRelayAddresses;
+    /**
+     * Fetch the current list of random masks from the server, while preserving local labels if present
+     */
+    async function refreshLocalLabelCache() {
+      /** @type {RandomMask[]} */
+      const relayAddresses = await apiRequest(apiRelayAddressesURL);
+      await browser.storage.local.set({ relayAddresses: applyLocalLabels(relayAddresses) });
+    }
+
+    /**
+     * Copy over locally-stored labels to the given addresses
+     *
+     * If the user has disabled server-side label storage, the add-on may still
+     * store data like labels and the websites an address has been used on
+     * locally. When we refresh the address data from the server, we'll want to
+     * copy over the labels from our local cache of addresses to the new list
+     * we just fetched.
+     *
+     * @param {RandomMask[]} addresses
+     * @returns {RandomMask[]}
+     */
+    function applyLocalLabels(addresses) {
+      if (siteStorageEnabled) {
+        return addresses;
+      }
+
+      const localAddressCache = browser.storage.local.get("relayAddresses").relayAddresses ?? [];
+      return addresses.map(address => {
+        const matchingLocalAddress = localAddressCache.find((localAddress) => {
+          return (
+            localAddress.id === address.id &&
+            localAddress.address === address.address &&
+            localAddress.domain === address.domain
+          );
+        });
+
+        return {
+          ...address,
+          description: matchingLocalAddress?.description ?? address.description,
+          generated_for: matchingLocalAddress?.generated_for ?? address.generated_for,
+          used_on: matchingLocalAddress?.used_on ?? address.used_on,
+        };
+      });
+    }
```

Comment:
Note: we'll probably want to call this function on [`masks` here](https://github.com/mozilla/fx-private-relay-add-on/blob/7f1653ec82a60cd35d68f4bfacf0fe7692a3c7a5/src/js/background/background.js#L63) as well, to solve https://github.com/mozilla/fx-private-relay-add-on/issues/331.